src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -3,203 +3,218 @@ namespace NServiceBus
+                data = sagaData;
+            }
+
+            static IContainSagaData DeepClone(IContainSagaData source)
+            {
+                var json = serializer.SerializeObject(source);
+                return (IContainSagaData)serializer.DeserializeObject(json, source.GetType());
+            }
+
+            public IContainSagaData GetSagaCopy()
+            {
+                return DeepClone(data);
+            }
+
+            public Entry UpdateTo(IContainSagaData sagaData)
+            {
+                return new Entry(sagaData, CorrelationId);
+            }
+        }
+
         public Task Complete(IContainSagaData sagaData, SynchronizedStorageSession session, ContextBag context)
         {
-            var inMemSession = (InMemorySynchronizedStorageSession) session;
-            inMemSession.Enlist(() =>
+            ((InMemorySynchronizedStorageSession)session).Enlist(() =>
             {
-                VersionedSagaEntity value;
-                if (data.TryRemove(sagaData.Id, out value))
+                var entry = GetEntry(sagaData, context);
+
+                if (sagas.TryRemoveConditionally(sagaData.Id, entry) == false)
                 {
-                    object lockToken;
-                    lockers.TryRemove(value.LockTokenKey, out lockToken);
+                    throw new Exception("Saga can't be completed as it was updated by another process.");
+                }
+
+                // saga removed
+                // clean the index
+                if (Equals(entry.CorrelationId, NoCorrelationId) == false)
+                {
+                    byCorrelationId.TryRemoveConditionally(entry.CorrelationId, sagaData.Id);
                 }
             });
-            return TaskEx.CompletedTask;
+
+            return Task.FromResult(0);
         }
 
-        public Task<TSagaData> Get<TSagaData>(string propertyName, object propertyValue, SynchronizedStorageSession session, ContextBag context) where TSagaData : IContainSagaData
+        public Task<TSagaData> Get<TSagaData>(Guid sagaId, SynchronizedStorageSession session, ContextBag context)
+            where TSagaData : IContainSagaData
         {
-            Guard.AgainstNull(nameof(propertyValue), propertyValue);
+            Entry value;
 
-            foreach (var entity in data.Values)
+            if (sagas.TryGetValue(sagaId, out value))
             {
-                if (!(entity.SagaData is TSagaData))
-                {
-                    continue;
-                }
+                context.Set(sagaId.ToString(), value);
 
-                var prop = typeof(TSagaData).GetProperty(propertyName);
-                if (prop == null)
-                {
-                    continue;
-                }
-                var existingValue = prop.GetValue(entity.SagaData);
+                var data = value.GetSagaCopy();
 
-                if (existingValue.ToString() != propertyValue.ToString())
+                if (data is TSagaData)
```

Comment:
Agree with @timbussmann and I cannot come up with a reason when this would be false. @andreasohlund any ideas?